### ISSUE TITLE
Strange interaction between `typing.Protocol` and unrelated `isinstance()` checks

### ISSUE THREAD

**Issue by AlexWaygood** [python/cpython#105144]
On `main`, an `isinstance()` check against `Sized` works just the same as it does on Python 3.11:

```pycon
>>> from typing import Sized
>>> isinstance(1, Sized)
False
```

However! If you first subclass `Sized` like this, `TypeError` is raised on that `isinstance()` check!

```pycon
>>> from typing import Sized, Protocol
>>> class Foo(Sized, Protocol): pass
...
>>> isinstance(1, Sized)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\alexw\coding\cpython\Lib\typing.py", line 1153, in __instancecheck__
    return self.__subclasscheck__(type(obj))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\alexw\coding\cpython\Lib\typing.py", line 1428, in __subclasscheck__
    return issubclass(cls, self.__origin__)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen abc>", line 123, in __subclasscheck__
  File "C:\Users\alexw\coding\cpython\Lib\typing.py", line 1793, in __subclasscheck__
    return super().__subclasscheck__(other)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen abc>", line 123, in __subclasscheck__
  File "C:\Users\alexw\coding\cpython\Lib\typing.py", line 1876, in _proto_hook
    raise TypeError("Instance and class checks can only be used with"
TypeError: Instance and class checks can only be used with @runtime_checkable protocols
```

This was originally reported by @vnmabus in https://github.com/python/typing_extensions/issues/207.

Note that (because of the `abc`-module cache), this _doesn't_ reproduce if you do an `isinstance()` check _before_ subclassing `typing.Sized`:

```pycon
>>> import typing
>>> isinstance(1, typing.Sized)
False
>>> class Foo(typing.Sized, typing.Protocol): pass
...
>>> isinstance(1, typing.Sized)
False
```

### Linked PRs
* gh-105152
* gh-105159
* gh-105160

**Comment by AlexWaygood** [comment]
This bisects to https://github.com/python/cpython/commit/b27fe67f3c643e174c3619b669228ef34b6d87ee

**Comment by AlexWaygood** [comment]
Also reproduces if you use `collections.abc.Sized` rather than `typing.Sized`:

```pycon
>>> import collections.abc
>>> import typing
>>> class Foo(collections.abc.Sized, typing.Protocol): pass
...
>>> isinstance(1, collections.abc.Sized)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\alexw\coding\cpython\Lib\abc.py", line 119, in __instancecheck__
    return _abc_instancecheck(cls, instance)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\alexw\coding\cpython\Lib\abc.py", line 123, in __subclasscheck__
    return _abc_subclasscheck(cls, subclass)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\alexw\coding\cpython\Lib\typing.py", line 1796, in __subclasscheck__
    return super().__subclasscheck__(other)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\alexw\coding\cpython\Lib\abc.py", line 123, in __subclasscheck__
    return _abc_subclasscheck(cls, subclass)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\alexw\coding\cpython\Lib\typing.py", line 1879, in _proto_hook
    raise TypeError("Instance and class checks can only be used with"
TypeError: Instance and class checks can only be used with @runtime_checkable protocols
```

**Comment by JelleZijlstra** [comment]
With some debugging we found that this is due to code in _abc.c that iterates over all the `__subclasses__` of the ABC:

https://github.com/python/cpython/blob/c05c31db8c9dfd708b9857bb57f8e5f3ce40266d/Modules/_abc.c#L768

Alex's PR #105152 works around this problem by re-arranging the code in typing.py. In #105159 I propose a fix on the ABC side instead. We should consider merging that PR too to prevent similar problems with other ABC subclasses.

**Comment by AlexWaygood** [comment]
This issue is fixed, and Jelle's decided not to pursue #105159, so I'll close this
